routes/users_routes.js

```diff
@@ -130,30 +130,43 @@ router.post('/lostpasswordreset', function(req, res) {
         User.findOne({
           email: req.body.email
         }, function(err, user) {
-          if (user) {
-            if (req.body.password2 != req.body.password) {
-              req.flash('danger', 'New passwords don\'t match');
-              res.render('user/passwordreset');
-            } else {
-              bcrypt.genSalt(10, function(err, salt) {
-                bcrypt.hash(req.body.password2, salt, function(err, hash) {
-                  if (err) {
-                    console.log(err);
-                  } else {
-                    user.password = hash;
-                    user.save(function(err) {
-                      if (err) {
-                        console.log(err)
-                        return;
-                      } else {
-                        req.flash('success', 'Password updated succesfully');
-                        return res.redirect('/user/login');
-                      }
-                    });
-                  }
-                });
+          if (err) {
+            console.error('Password reset find user error:', err);
+            res.sendStatus(500);
+            return;
+          }
+          if (!user) {
+            req.flash('danger', 'No user with that email found! Are you sure you created an account?');
+            res.render('user/passwordreset');
+          }
+          if (req.body.password2 != req.body.password) {
```

Comment:
Because you are not using else, you can send multiple responses and crash the server in the case of multiple errors.

Remove 'return' and refactor this such that it deterministically sends exactly one response

Reply:
Good catch, fixed. This function would turn into a giant tree without using return so I'd much prefer to leave that in.

Reply:
Be careful about using return with nested callbacks, it could just return the anonymous function, and still result in multiple responses. This looks fine to me though.